homeassistant/components/azure_data_explorer/__init__.py

```diff
@@ -0,0 +1,230 @@
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        entities_filter: vol.Schema,
+    ) -> None:
+        """Initialize the listener."""
+
+        self.hass = hass
+        self._entry = entry
+        self._entities_filter = entities_filter
+
+        self._client = AzureDataExplorerClient(**self._entry.data)
+        self._send_interval = self._entry.options[CONF_SEND_INTERVAL]
+        self._max_delay = self._entry.options.get(CONF_MAX_DELAY, DEFAULT_MAX_DELAY)
+
+        self._shutdown = False
+        self._queue: asyncio.PriorityQueue[  # pylint: disable=unsubscriptable-object
+            tuple[int, tuple[datetime, State | None]]
+        ] = asyncio.PriorityQueue()
+        self._listener_remover: Callable[[], None] | None = None
+        self._next_send_remover: Callable[[], None] | None = None
+
+    async def async_start(self) -> None:
+        """Start the component.
+
+        This register the listener and
+        schedules the first send.
+        """
+
+        self._listener_remover = self.hass.bus.async_listen(
+            MATCH_ALL, self.async_listen
+        )
+        self._schedule_next_send()
+
+    async def async_stop(self) -> None:
+        """Shut down the ADX by queueing None, calling send, join queue."""
+        if self._next_send_remover:
+            self._next_send_remover()
+        if self._listener_remover:
+            self._listener_remover()
+        await self._queue.put(
+            (3, (utcnow(), None))
+        )  # <---WHY IS THIS NEEDED!!! Creates error when sending none
+        await self.async_send(None)
+        await self._queue.join()
+
+    def update_options(self, new_options: dict[str, Any]) -> None:
+        """Update options."""
+        self._send_interval = new_options[CONF_SEND_INTERVAL]
+
+    async def test_connection(self) -> None:
+        """Test the connection to the Azure Data Explorer service."""
+        await self.hass.async_add_executor_job(self._client.test_connection)
+
+        return None
+
+    def _schedule_next_send(self) -> None:
+        """Schedule the next send."""
+        if not self._shutdown:
+            self._next_send_remover = async_call_later(
+                self.hass, self._send_interval, self.async_send
+            )
+
+    async def async_listen(self, event: Event) -> None:
+        """Listen for new messages on the bus and queue them for AEH."""
+        if state := event.data.get("new_state"):
+            await self._queue.put((2, (event.time_fired, state)))
+
+    async def async_send(self, _) -> None:
+        """Write preprocessed events to Azure Data Explorer."""
+
+        adx_events = ""
+        while not self._queue.empty():
+            _, event = self._queue.get_nowait()
+            dropped = 0
+            adx_event, dropped = self._parse_event(*event, dropped)
+            if dropped == 0:
+                adx_events += str(adx_event)
```

Comment:
adx_event is already a string, and with a filter or None this `str` is not needed